redis/src/caching/cache_manager.rs

```diff
@@ -0,0 +1,226 @@
+use crate::caching::statistics::Statistics;
+use crate::caching::{CacheConfig, CacheStatistics};
+use crate::cmd::{CommandCacheInformation, CommandCacheInformationByRef};
+use crate::Value;
+use lru::LruCache;
+use std::cmp::min;
+use std::collections::hash_map::DefaultHasher;
+use std::hash::Hasher;
+use std::num::NonZeroUsize;
+use std::ops::Add;
+use std::sync::Arc;
+use std::time::{Duration, Instant};
+use tokio::sync::watch::{channel, Receiver, Sender};
+
+type RedisCmd = Vec<u8>;
+type RedisKey = Vec<u8>;
+type LRUCacheShard = LruCache<RedisKey, CacheItem>;
+
+struct CacheCmdEntry {
+    cmd: RedisCmd,
+    value: Option<Value>,
+    receiver: Receiver<Value>,
```

Comment:
what is this? what purpose does it serve?

Reply:
If multiple request is trying to access same key with same command, only one request will send request to server, other requests will wait this  `receiver`
```
if receiver.changed().await.is_ok() {
                self.statistics.increase_hit(1);
                return Ok(Some(receiver.borrow().clone()));
           }
```